tests/e2e/postgres/locking_test.go

```diff
@@ -0,0 +1,227 @@
+package postgres_test
+
+import (
+	"context"
+	"errors"
+	"sort"
+	"sync"
+	"testing"
+
+	"github.com/pressly/goose/v4"
+	"github.com/pressly/goose/v4/internal/check"
+	"golang.org/x/sync/errgroup"
+)
+
+func TestLockModeAdvisorySession(t *testing.T) {
+	t.Parallel()
+
+	// The migrations are written in such a way that they cannot be applied concurrently. This test
+	// ensures that the advisory session lock mode works as expected.
+
+	options := goose.DefaultOptions().
+		SetDir(migrationsDir).
+		SetVerbose(testing.Verbose()).
+		SetLockMode(goose.LockModeAdvisorySession) /* ---------------- advisory session lock mode */
```

Comment:
I believe this satisfies the most common use cases raised in #335 by @tomwhipple @jessedearing @dahu33

(I still, need to clean up the internal abstractions, but tests are passing and things are shaping up).